### ISSUE TITLE
Work Allocator and Worker Threads.

### ISSUE THREAD

**Issue by da2ce7** [josecelano/library-consumer#2]
Consider using a more advanced approach to the locking.

### Work Allocator  

Runs On:

- All New Comment on Library Main.
- All New Comment on Library Consumer Main.

Checks:

1. All "Library Update" locks are released on Library Consumer Main.
2. An unprocessed Comment on Library Main.

If All Checks Pass Create Empty Commit: 

Commit Message (template):

```
CLAIM LOCK: JOB: Library Update [library-aaa]

Update Libary Submodule and Process Changes:
https://github.com/josecelano/library-aaa/commit/0cdc89d925e48d185bde7376949ccca39579c82f
->
https://github.com/josecelano/library-aaa/commit/ea918cb0940147200b477a4a97e417d0799644d0
```

**Fast-Forward Merge into Main Branch.**

### Library Update Worker Thread

Runs On:

- All New Comment on Library Consumer Main.
- As a cron job.

Checks:

1. Singe Claimed Lock of type: "Library Update"

Preforms:

- Create commit with Update submodule according to the Lock Message.
- Create commits with other related processing.

If All Work is successfully completed Create Empty Commit: 

```
RELEASE LOCK: JOB DONE: Library Update [library-aaa]

Successfully completed Library Update according to specification in Lock:
https://github.com/josecelano/library-consumer/commit/9aebfe3b5cd5a559933014b0c8572b0c4a20453e

```

**Fast-Forward Merge All Commits, (the work and the Lock Release), into Main Branch.**

or

**Open Pull Request for Human or Automated Review.**

---

**Please note 1**: Either all the work and the lock release is merged in automatically ... if there is an error, the lock stays open and waits for human intervention.  - We do not want the work to be merged in, but the lock not released.

**Please note 2**: The locks are now typed. So that there can be potentially multiple threads working out of order. Work on the website CSS may be on another thread and would be ignored by this library update process.

- If another (unrelated) worker thread was merged in-between, the fast-forward merge should fail, and the worker thread would be restated automatically by the cron job. (and attempted again).

**Please note 3**: We can have Global Locks that would be human controlled, that would effectively pause the automated workflows.

**Comment by josecelano** [comment]
I did:

- A branch for the issue
- Open a PR
- There was a syntax error in the workflow
- I fixed
- I was not able to run the fixed workflow
- I deleted the workflow using the GitHub interface
- I was not able to see the workflow again on the interface
- I had to merge the PR into main
- Now I'm working directly on the main branch with the new version

**Comment by josecelano** [comment]
THe script to parse the commit body is working on my dev env:

```
josecelano@josecelano:~/Documents/github/josecelano/library-consumer$ bash ./scripts/get-active-lock.sh 
* 3fd9fb6565d9acdfa6960e70747fb316d64318e3 - CLAIM LOCK: JOB: Library Update [library-aaa]
Active lock found in commit 3fd9fb6565d9acdfa6960e70747fb316d64318e3
COMMIT LINE:     Update Library Submodule and Process Changes:
COMMIT LINE:     https://github.com/josecelano/library-aaa/commit/ecc0501379fbf9590fa09f3f139a632fa03837bf
COMMIT LINE:     ->
COMMIT LINE:     https://github.com/josecelano/library-aaa/commit/f35514303a93ae247cf44ab767aa8d59a368ec28
COMMIT LINE:     PREVIOUS_REF=ecc0501379fbf9590fa09f3f139a632fa03837bf
COMMIT LINE:     CURRENT_REF=https://github.com/josecelano/library-aaa/commit/f35514303a93ae247cf44ab767aa8d59a368ec28
::set-output name=previous_ref::ecc0501379fbf9590fa09f3f139a632fa03837bf
::set-output name=current_ref::https://github.com/josecelano/library-aaa/commit/f35514303a93ae247cf44ab767aa8d59a368ec28
::set-output name=active_lock::true
```

but for some reason, it's not working [on the runner](https://github.com/josecelano/library-consumer/runs/4844692493?check_suite_focus=true#step:4:4).

**Comment by josecelano** [comment]
It seems it's almost workgin now:

```
    CLAIM LOCK: JOB: Library Update [library-aaa]\n \n Update Library Submodule and Process Changes:\n https://github.com/josecelano/library-aaa/commit/f35514303a93ae247cf44ab767aa8d59a368ec28\n ->\n https://github.com/josecelano/library-aaa/commit/92dda99e546437300ad5a6d2fe2a7f70e4bad2ca\n \n PREVIOUS_REF=f35514303a93ae247cf44ab767aa8d59a368ec28\n \n CURRENT_REF=https://github.com/josecelano/library-aaa/commit/92dda99e546437300ad5a6d2fe2a7f70e4bad2ca\n
```

Line breaks in the commit message are inserted literally as `\n`.

I have to find out how to assign multiline string values in a step command (inline)

**Comment by josecelano** [comment]
I could fix the line breaks problem with:

```
- name: Release lock
  if: ${{ steps.get-active-lock.outputs.active_lock == 'true' }}
  run: |
    COMMIT_MESSAGE="RELEASE LOCK: JOB DONE: Library Update [library-aaa]
    \n
    Successfully completed Library Update according to specification in Lock:
    https://github.com/josecelano/library-aaa/commit/${{ steps.get-active-lock.outputs.current_ref }}"
    git commit --allow-empty -m "$(echo -e "$COMMIT_MESSAGE")"
```

The problem was in the git commit command.These are the claim and release lock commits:

Claim commit:
```
josecelano@josecelano:~/Documents/github/josecelano/library-consumer$ git show 6d1b72d
commit 6d1b72d40669bc7e2a4c5ec903040c843fc3e987
Author: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Date:   Tue Jan 18 11:05:22 2022 +0000

    CLAIM LOCK: JOB: Library Update [library-aaa]
    
    Update Library Submodule and Process Changes:
    https://github.com/josecelano/library-aaa/commit/f35514303a93ae247cf44ab767aa8d59a368ec28
    ->
    https://github.com/josecelano/library-aaa/commit/845e5ab87099e30c017851c64304c4b8d6625268
    
    PREVIOUS_REF=f35514303a93ae247cf44ab767aa8d59a368ec28
    CURRENT_REF=845e5ab87099e30c017851c64304c4b8d6625268
```

Release commit:
```
josecelano@josecelano:~/Documents/github/josecelano/library-consumer$ git show af8c66c
commit af8c66cdf8b06f3406afbb6384101b212c9c0505 (HEAD -> main, origin/main)
Author: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Date:   Tue Jan 18 11:06:26 2022 +0000

    RELEASE LOCK: JOB DONE: Library Update [library-aaa]
    
    Successfully completed Library Update according to specification in Lock:
    https://github.com/josecelano/library-aaa/commit/845e5ab87099e30c017851c64304c4b8d6625268
```

**Comment by josecelano** [comment]
hi @da2ce7 I think I have finished the implementation of this new version. I have also updated the README. A couple of comments:

1. I think the main drawback of this solution could be "deadlocks" if we have a high rate of merges. Please read the README for a full list of pros and cons.
2. I have also added some lines at the end of the claim commit message in other to make it easier to parse the commit sha with bash.
```
PREVIOUS_REF=f35514303a93ae247cf44ab767aa8d59a368ec28
CURRENT_REF=845e5ab87099e30c017851c64304c4b8d6625268
```
Basically, all lines starting with `XX..XX=` are considered an input variable for the worker. Here we can follow hundreds of options: json, yml, toml, GPG colons format, GitHub actions format, etcetera. Let me know which one of them you consider the best in this case. It could be also defined by each worker. If we continue using the one I have implemented maybe we should add at least a mark for `BEGIN-VARIABLES` section.
3. I think this solution it's the "optimistic locking" solution. The previous one was the "pessimistic locking".
4. There is one thing in your proposal I have not added: trigger the workflow on the push event on the library repo. The way to do that is by adding a workflow to the library. We can do it but I was trying to avoid coupling between the library and the consumer. The main problem is we can only update every 10 minutes, but I don't think that's a problem right now for the project where we are planning to add this solution.

**Comment by da2ce7** [comment]
Hello @josecelano 

Good Work.

Some comments:

In: https://github.com/josecelano/library-consumer/blob/main/.github/workflows/work-allocator.yml

`./scripts/check-lock.sh` should be run as the first step, no need to continue workflow if there is a already lock claimed.

---

In: https://github.com/josecelano/library-consumer/blob/main/scripts/check-lock.sh

This script should go throughout the entire git-history on this branch, and should match locks with their releasing pairs.
Just checking for the most recent lock open message is not good practice.

Maybe a program called `get-active-locks.ts` would be a better replacement.  (In general I would prefer to avoid using shell scripts).  Type Script or JavaScript would be a better replacement.

`get-active-locks.ts` Would return an array of active locks, with their git commit-hashes and scope.  If there are no active locks, the array can return empty.

---

> We force fast forward merges for all PRs.

This is not the case, we can use normal merges for all PR's that:
- Do not claim or release a lock.

---

> If we want to avoid too many merge conflicts, we have to pause other manual PR merges when the workflow is being executed.

If the manual PR dose not claim or release a lock it can safely merged at any time.

- The Automated workflows that release a lock will fail when they cannot fast-forward merge, but this is not a problem because they will be automaticity retried by the cron job.

---

> We only can have one global lock at the time. Actually, this solution provides a way to lock any other merge into the main branch.

With this worker approach we are having limited the scope of the locks claimed.  It is indeed possible to construct a special "Global" or "World" lock, however we would only use this for special occasions.

---

> With the optimistic approach we can waste a lot of CPU (or other resources) re-processing workflows. If we have too many parallel tasks, merge conflicts could even lead to a deadlock.

I do not believe that this the case.  Workflows fail early when they cannot secure their lock.

- The only common case where there would be reprocessed work is when a unrelated PR is merged in in-between the processing time of a Worker Thread.  -  This would require the worker thread to be attempted again later after the failed fast-forward merge.

- Unless we do something very wrong, there should never be a deadlock from merge conflicts with the "optimistic approach", When there is a merge conflict, the work is simply disregarded and re-attempted latter on by the cron job, therefore releasing the lock at a later time (without the merge conflict).

**Comment by josecelano** [comment]
1. `/scripts/check-lock.sh` first step

You are right. That's because in my previous version I was creating and releasing the lock on every workflow executing and I was trying to avoid creating the lock if it was not needed because there were no updates in the library. I will change it.

2. Better way to extract the current claimed lock list

I totally agree with you. I implemented that version because I thought it was the most basic implementation. In fact, I was looking for the most basic implementation because I planned to use bash and for me, it's a pain (probably because I do not know it very well). So both things are related. For some reason, I tend to prefer the `bash` for proof of concepts instead of Python. I think both options have their pros and cons. For me, the cons of Python are the setup and how to deal with console commands. That's why I probably tend to use bash for the first approach. TypeScript could be an option but since we are not using it in other projects I did not want to introduce one more language.

Do you think we could create an embed GitHub action in TypeScript to handle the lock system? Maybe we just need to wrap the bash scripts. My idea was to extract those actions after finishing the first implementation because I thought that way it would be clearer to me what I should extract.

3. We force fast forward merges for all PRs.

I meant if we want to enforce fast forward merges for the main branch I think the only way to do that is enabling the option "Require linear history" in the repo setting. And as far as I understand that forces all merges into main to be with rebase. But I can be wrong.

4. If the manual PR does not claim or release a lock it can safely merge at any time

Yes, that's true. I was thinking about how to avoid wasting execution time, but as you say in the last point, It does not matter, the workflow will quit almost immediately. At least other threads of the same workflow.

5. global lock

I was wrong, even my previous solution could work with more than one lock. Again, I was obfuscated with trying to avoid running other workflows in parallel that are going to fail when merging. But that does not mean that is not possible to have more than one lock.

6. deadlock

Yes, I suppose we have to introduce some kind of dependency between workflows in order to produce a deadlock. But with a lot of branches pending to merge, we could have some workflows waiting too long. But I suppose it's very hard to try to predict all the potential problems.

Maybe we can define some rules like: do not generate side effects during a workflow execution if you need to revert them when if the workflow has to be re-executed, etcetera.

Finally,

Next steps:

- I will update the README with this discussion.
- I will apply your proposed changes (re-order step)
- I will extract scripts into TypeScript GitHub actions. Maybe we can re-use those actions later in the website project, What do you think? I think the locking system can be kept decoupled from the rest of the Python code, but I'm not sure yet.
- Have you thought about an alternative format for the commit message?

cc @yeraydavidrodriguez

**Comment by da2ce7** [comment]
Hello @josecelano

> Do you think we could create an embed GitHub action in TypeScript to handle the lock system? Maybe we just need to wrap the bash scripts. My idea was to extract those actions after finishing the first implementation because I thought that way it would be clearer to me what I should extract.

I think that this is the perfect case to use/make a TypeScript Action.

---

> 3. We force fast forward merges for all PRs.
>
> I meant if we want to enforce fast forward merges for the main branch I think the only way to do that is enabling the option "Require linear history" in the repo setting. And as far as I understand that forces all merges into main to be with rebase. But I can be wrong.

`git push` by default will not allow for a non-fast-forward upstream update.  We can add the `--atomic` option to ask the server to make the push transaction atomic (if supported).

So we don't really need to force the `Require Linear History` option in the branch protection rules.

---

> Have you thought about an alternative format for the commit message?

I think that we should have a brainstorm around the vocabulary and messaging, including what formats we use for encoding the data, etc, that we use for the commit title and message body.  Overall I'm not really satisfied in calling everything a `LOCK`, I think that we have a work-allocator and thread-workers.

cc: @yeraydavidrodriguez

**Comment by josecelano** [comment]
Hi @da2ce7 

OK, I'm going to move the code to a TS GitHub embedded action. Maybe the [GPG import action](https://github.com/crazy-max/ghaction-import-gpg) we are already using could be a good starting point.

It seems [atomic pushes are allowed](https://github.blog/2015-04-30-git-2-4-atomic-pushes-push-to-deploy-and-more/), but I do not see the point here. I think we only push the local `main` branch into the remote `main` branch (we do not have multiple pushes). So it should be enough if we continue with the default push behaviour. As you said we can still allow other merges, we do not need to force linear history. I was confused.

OK, I'm going to create a new issue to discuss the best format for the commit messages. I have the same impression. For me, it's more like a message queue with events (using a lock mechanism to avoid creating duplicate events and processing them twice). I think we have a kind of N message queues each of them accepting only one message.

1. We receive the infrastructure event (the library has been updated)
2. We save the event (`library_aaa_was_updated`) in the event store (git database, commit object). We need to be sure we do not create duplicate events since we have parallel threads listening to Github events (schedule workflow that can be overlapped).
3. We have workers that process the events and we need an ACK system (another event) to indicate the events were processed.
In my previous version I wanted to created a json object representing the event inside the commit message:

```
{
  "name": "library_was_udpated",
  "version": 1,
  "payload": {
     "previous_ref": "...",
     "current_ref": "...",
     "submodule": "...",
     "submoudle_url": "...",
  },
  "metadata": {
     "runner_id": "..",
     "created_at": "..."
  }
}
```

Since we use json to pass data around we could use it in this case. We can mix it with a more human-friendly short description like the one we have now.

**Comment by josecelano** [comment]
hi @da2ce7 I have updated the README and I have also moded the `/scripts/check-lock.sh` script to the top. In fact, I realized I do not need it anymore because It does the same as `get-active-lock.sh`. [Here](https://github.com/josecelano/library-consumer/pull/6) is the [PR](https://github.com/josecelano/library-consumer/pull/6).

I have also created two new issues:

- [Brainstorm around the vocabulary and messaging](https://github.com/josecelano/library-consumer/issues/4)
- [Embedded Typescript actions to implement the key workflow steps](https://github.com/josecelano/library-consumer/issues/7)
